superset/db_engine_specs/drill.py

```diff
@@ -117,3 +120,23 @@ def get_url_for_impersonation(
                 )
 
         return url
+
+    @classmethod
+    def fetch_data(
+        cls,
+        cursor: Any,
+        limit: int | None = None,
+    ) -> list[tuple[Any, ...]]:
+        """
+        Custom `fetch_data` for Drill.
+
+        When no rows are returned, Drill raises a `RuntimeError` with the message
```

Comment:
@betodealmeida is the Drill SQLAlchemy dialect or DB-API not conforming to the standard? Ideally it would be great to adopt the "shift left" approach and ensure this is handled correctly upstream as opposed for us having to write custom logic.

Reply:
@john-bodley I'm one of the authors of the Drill SQLAlchemy dialect.  It's been a few years however, so I don't remember much about how it works.  If something should be fixed there, I'm happy to do it.  However, could someone please explain what the exact issue is?

Reply:
@cgivre looking at [PEP-249](https://peps.python.org/pep-0249/) it seems like a `StopIteration` should be raised if `Cursor.next()` is called and the result set is exhausted. In the base engine spec the [`fetch_data()`](https://github.com/apache/superset/blob/a912fafb6d592703c5b66f59be180220329ab9c4/superset/db_engine_specs/base.py#L749-L787) method calls `Cursor.fetchall()` or `Cursor.fetchmany()` rather than `Cursor.next()`. It would be prudent to double check the Drill DB-API and SQLAlchemy dialects are behaving in accordance with the standards.

I guess the TL;DR is I'm not sure why we need this custom logic for Drill.

Reply:
@john-bodley we believe this is now fixed sqlalchemy-drill. That's certainly where it _should_ be fixed.